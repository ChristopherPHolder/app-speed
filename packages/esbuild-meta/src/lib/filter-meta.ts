import { CommandModule, CommandBuilder, Options, InferredOptionTypes, Argv } from 'yargs';
import { Metafile } from 'esbuild';
import {
  extractEntryPoints,
  filterMetaFromEntryPoints,
  removeDynamicImports,
  getJson,
  makeJson,
} from './utils.js';

export const DEMAND_STATS_PATH = 'The path to a stats.json file is required';

const statsPath = {
  alias: 's',
  type: 'string',
  description: 'The path to the stats.json file',
  demandOption: DEMAND_STATS_PATH,
} as const satisfies Options;

const outPath = {
  alias: 'o',
  type: 'string',
  description: 'The path where the new file should be saved',
  default: 'initial-stats.json'
} as const satisfies Options;

const excludeDynamicImports = {
  alias: 'eDI',
  type: 'boolean',
  description: 'Should the dynamic imports be filtered out of the output chunk imports',
  default: false,
} as const satisfies Options;

const entryPoints = {
  alias: 'e',
  type: 'array',
  default: ['main-', 'polyfills-'],
  coerce: (i) => i.map((value: never) => String(value)),
  description: 'Entry points that should be considered for the bundle',
} as const satisfies Options;

const filterMetaOptions = { statsPath, outPath, excludeDynamicImports, entryPoints };

type FilterMetaOptions = InferredOptionTypes<typeof filterMetaOptions>;
type FilterMetaCommandModule = CommandModule<unknown, FilterMetaOptions>;

const filterMetaBuilder: CommandBuilder<unknown, FilterMetaOptions> = (argv: Argv<unknown>) => {
  return argv.options(filterMetaOptions);
}

const filterMetaHandler: FilterMetaCommandModule['handler'] = (argv: FilterMetaOptions) => {
  const meta = getJson<Metafile>(argv.statsPath);
  const entryPoints = extractEntryPoints(meta, argv.entryPoints);
  filterMetaFromEntryPoints(meta, entryPoints);
  if (argv.excludeDynamicImports) {
    removeDynamicImports(meta);
  }
  makeJson(argv.outPath, meta);
  console.log('Filtered Meta File was successfully created as ' + argv.outPath);
}

export const filterMetaCommand: FilterMetaCommandModule = {
  command: 'filter',
  describe: 'Filters the meta file to only include chunks required by specified entry points',
  aliases: 'f',
  builder: filterMetaBuilder,
  handler: filterMetaHandler
}
