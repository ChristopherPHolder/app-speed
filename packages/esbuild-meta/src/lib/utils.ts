import { readFileSync, writeFileSync } from 'node:fs';
import { join, normalize } from 'node:path';
import { Metafile } from 'esbuild';

export const INVALID_FILE_PATH_ERROR_MSG = (path: string) => `No file found at ${path}`;

export function getJson<T = any>(path: string) {
  const normalizedPath = normalize(path);
  try {
    return JSON.parse(readFileSync(normalizedPath, {encoding: 'utf-8'})) as T;
  }
  catch (e) {
    throw new Error(INVALID_FILE_PATH_ERROR_MSG(normalizedPath));
  }
}

export function makeJson(path: string, file: any) {
  writeFileSync(join(...[path]), JSON.stringify(file, null, 4), {encoding: 'utf-8'});
}

export function removeDynamicImports(meta: Metafile): void {
  const chunksPaths = Object.keys(meta.outputs);
  for (const chunkPath of chunksPaths) {
    const imports = meta.outputs[chunkPath].imports;
    for (const importIndex in imports) {
      if (imports[importIndex].kind === "dynamic-import") {
        delete meta.outputs[chunkPath].imports[importIndex];
      }
    }
  }
}

const matchesPattern = (fileName: string, patterns: string[]) => patterns.some(substring => fileName.includes(substring));

export function extractEntryPoints(meta: Metafile, patterns: string[]): string[] {
  return Object.keys(meta.outputs).filter(fileName => matchesPattern(fileName, patterns));
}

export function filterMetaFromEntryPoints(meta: Metafile, entryPoints: string[]) {
  const extractedChunks = new Set<string>();
  const alreadyExtractedChildren = (chunk: string) => extractedChunks.has(chunk);
  const addToExtractedChunks = (chunk: string) => extractedChunks.add(chunk) && extractedChunks.add(`${chunk}.map`);
  const childImportedChunks = (chunk: string) => meta['outputs'][chunk]['imports'].filter(({kind}) => kind === "import-statement" || "dynamic-import").map(({path}) => path);
  function extractImport(chunks: string[]) {
    for (const chunk of chunks) {
      if (alreadyExtractedChildren(chunk)) continue;
      addToExtractedChunks(chunk);
      extractImport(childImportedChunks(chunk));
    }
  }
  function removeNotReferencedChunks() {
    const outputChunks = Object.keys(meta['outputs']);
    for (const chunk of outputChunks) {
      if (extractedChunks.has(chunk)) continue;
      delete meta['outputs'][chunk];
    }
  }

  extractImport(entryPoints);
  removeNotReferencedChunks();
}
