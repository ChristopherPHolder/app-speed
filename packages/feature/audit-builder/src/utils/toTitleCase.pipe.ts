import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'toTitleCase',
  standalone: true,
})
export class ToTitleCasePipe implements PipeTransform {
  transform(value: string): unknown {
    return value
      .replace(/([A-Z])/g, ' $1') // Add space before capital letters
      .replace(/^\w|[A-Z]|\b\w/g, (word) => word.toUpperCase()); // Capitalize first character of each word
  }
}
