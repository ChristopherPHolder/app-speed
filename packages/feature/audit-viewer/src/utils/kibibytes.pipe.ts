import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'kibibytes',
  standalone: true,
})
export class KibibytesPipe implements PipeTransform {
  transform(bytes: number): string {
    return Math.round(bytes / 1024) + ' KiB';
  }
}
