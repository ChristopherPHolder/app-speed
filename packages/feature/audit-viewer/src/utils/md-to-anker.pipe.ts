import { inject, Pipe, PipeTransform } from '@angular/core';
import { DomSanitizer, SafeHtml } from '@angular/platform-browser';

@Pipe({
  name: 'mdToAnker',
  standalone: true,
  pure: false
})
export class MdToAnkerPipe implements PipeTransform {
  private readonly sanitizer = inject(DomSanitizer);
  transform(value: string): SafeHtml {
    return this.sanitizer.bypassSecurityTrustHtml(value.replaceAll(
      /(\[([^\]]+)])\(([^)]+)\)/g,
      '<a target="_blank" href="$3">$2</a>'
    ));
  }
}
