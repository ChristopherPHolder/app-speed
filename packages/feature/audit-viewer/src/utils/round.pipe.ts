import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'round',
  standalone: true,
})
export class RoundPipe implements PipeTransform {
  transform(value: number): number {
    return Math.round(value)
  }
}
