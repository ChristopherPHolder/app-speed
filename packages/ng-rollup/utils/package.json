{
  "name": "@app-speed/ng-rollup/utils",
  "version": "0.0.1",
  "description": "Utility library for using rollup as a post processor for an angular app",
  "main": "./index.cjs",
  "module": "index.js",
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.js"
    }
  },
  "dependencies": {},
  "keywords": ["post-processor", "angular", "rollup", "esbuild"],
  "author": "Christopher Holder"
}
