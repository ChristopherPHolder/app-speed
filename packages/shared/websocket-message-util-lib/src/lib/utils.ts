import { CONDUCTOR_STAGE } from './constants';
import { StageChangeMessage, ConductorStages } from './types';

function isStageChangeMessage(message: unknown): message is StageChangeMessage {
  return (
    typeof message === 'object' &&
    message !== null &&
    'type' in message &&
    'stage' in message &&
    message.type === 'stage-change'
  );
}

function isDoneStageChangeMessage(message: unknown): message is StageChangeMessage<ConductorStages['DONE']> {
  return isStageChangeMessage(message) && message.stage === CONDUCTOR_STAGE.DONE;
}

export { isStageChangeMessage, isDoneStageChangeMessage };
