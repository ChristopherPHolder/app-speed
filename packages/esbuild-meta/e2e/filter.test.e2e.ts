import { describe, expect, it } from 'vitest';
import { cliProcess } from './utils.js';
import { DEMAND_STATS_PATH } from '../src/lib/filter-meta.js';
import { INVALID_FILE_PATH_ERROR_MSG } from '../src/lib/utils.js';

describe('filter command', () => {
  it('should have a help option', async () => {
    const { stdout, stderr, code } = await cliProcess('esbuild-meta filter --help');
    expect(stdout).toMatchSnapshot();
    expect(stderr).toBeFalsy();
    expect(code).toBe(0);
  });

  it('should demand stats path option', async () => {
    const { stdout, stderr, code } = await cliProcess('esbuild-meta filter');
    expect(stderr).toContain(DEMAND_STATS_PATH);
    expect(stdout).toBeFalsy();
    expect(code).toBe(1);
  });

  it('should throw if stats path does not point to a file', async () => {
    const INVALID_STATS_FILE = 'invalid-path.json';
    const { stdout, stderr, code } = await cliProcess(`esbuild-meta filter --statsPath ${INVALID_STATS_FILE}`);
    expect(stderr).toContain(INVALID_FILE_PATH_ERROR_MSG('invalid-path.json'));
    expect(stdout).toBeFalsy();
    expect(code).toBe(1);
  });
});
