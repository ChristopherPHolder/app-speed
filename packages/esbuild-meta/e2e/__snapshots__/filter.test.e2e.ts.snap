// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter command > should have a help option 1`] = `
"esbuild-meta filter

Filters the meta file to only include chunks required by specified entry points

Options:
  -s, --statsPath                     The path to the stats.json file  [string] [required]
  -o, --outPath                       The path where the new file should be saved  [string] [default: "initial-stats.json"]
      --excludeDynamicImports, --eDI  Should the dynamic imports be filtered out of the output chunk imports  [boolean] [default: false]
  -e, --entryPoints                   Entry points that should be considered for the bundle  [array] [default: ["main-","polyfills-"]]
  -v, --version                       Show version number  [boolean]
  -h, --help                          Show help  [boolean]
"
`;
